vsintegration/src/FSharp.Editor/Common/CodeAnalysisExtensions.fs

```diff
@@ -0,0 +1,108 @@
+﻿[<AutoOpen>]
+module internal Microsoft.VisualStudio.FSharp.Editor.CodeAnalysisExtensions
+
+open Microsoft.CodeAnalysis
+open Microsoft.FSharp.Compiler.Range
+
+type Project with
+
+    /// Returns the projectIds of all projects within the same solution that directly reference this project
+    member this.GetDependentProjectIds () =
+        this.Solution.GetProjectDependencyGraph().GetProjectsThatDirectlyDependOnThisProject this.Id
+
+
+    /// Returns all projects within the same solution that directly reference this project.
+    member this.GetDependentProjects () =
+        this.Solution.GetProjectDependencyGraph().GetProjectsThatDirectlyDependOnThisProject this.Id
+        |> Seq.map this.Solution.GetProject
+
+
+    /// Returns the ProjectIds of all of the projects that this project directly or transitively depneds on
+    member this.GetProjectIdsOfAllProjectsThisProjectDependsOn () =
+        let graph = this.Solution.GetProjectDependencyGraph()
+        let transitiveDependencies = graph.GetProjectsThatThisProjectTransitivelyDependsOn this.Id
+        let directDependencies = graph.GetProjectsThatThisProjectDirectlyDependsOn this.Id
+        Seq.append directDependencies transitiveDependencies
+        
+
+    /// The list all of the projects that this project directly or transitively depneds on
+    member this.GetAllProjectsThisProjectDependsOn () =
+        this.GetProjectIdsOfAllProjectsThisProjectDependsOn ()
+        |> Seq.map this.Solution.GetProject
+
+
+type Solution with 
+
+    /// Try to get a document inside the solution using the document's name
+    member self.TryGetDocumentNamed docName =
+        self.Projects |> Seq.tryPick (fun proj ->
+            proj.Documents |> Seq.tryFind (fun doc -> doc.Name = docName))
+
+
+     /// Try to find the documentId corresponding to the provided filepath within this solution  
+    member self.TryGetDocumentFromPath filePath =
+        self.GetDocumentIdsWithFilePath filePath
```

Comment:
This will sometimes fail because FSharp often returns file paths like e.g. `C:\visualfsharp\src\fsharp\FSharp.LanguageService.Compiler\..\vs\service.fsi`
I did normalization with GetFullPathSafe to work around.